projectorganizer/src/prjorg-sidebar.c

```diff
@@ -300,6 +300,199 @@ static void on_follow_active(GtkToggleToolButton *button, G_GNUC_UNUSED gpointer
+{
+	GtkTreeSelection *treesel = NULL;
+	GtkTreeModel *model = NULL;
+	GtkTreeIter iter;
+
+	gchar *locale_path = NULL;
+
+	if (prj_org)
+		treesel = gtk_tree_view_get_selection(GTK_TREE_VIEW(s_file_view));
+
+	/* try to get path from tree selection */
+	if (treesel && gtk_tree_selection_get_selected(treesel, &model, &iter))
+	{
+		if (gtk_tree_model_iter_has_child(model, &iter))
+		{
+			gchar *utf8_path;
+			GtkTreeView *tree_view;
+			GtkTreePath *tree_path;
+
+			tree_view = GTK_TREE_VIEW(s_file_view);
+			tree_path = gtk_tree_model_get_path (model, &iter);
+
+			utf8_path = build_path(&iter);
+			locale_path = utils_get_locale_from_utf8(utf8_path);
+			g_free(utf8_path);
+		}
+		else
+		{
+			GIcon *icon;
+			gtk_tree_model_get(model, &iter, FILEVIEW_COLUMN_ICON, &icon, -1);
+			if (icon)
+			{
+				gchar *utf8_path;
+				utf8_path = build_path(&iter);
+				locale_path = utils_get_locale_from_utf8(utf8_path);
+
+				g_free(utf8_path);
+				g_object_unref(icon);
+			}
+		}
+	}
+
+	/* Make sure path exists */
+	if(locale_path && !g_file_test(locale_path, G_FILE_TEST_EXISTS))
+	{
+		g_free(locale_path);
+		locale_path = NULL;
+	}
+
+	/* try to get path from document */
+	if (!locale_path)
+	{
+		GeanyDocument *doc = NULL;
+		doc = document_get_current();
+		if (DOC_VALID(doc))
+			locale_path = g_strdup(doc->real_path);
+	}
+
+	/* Make sure path exists */
+	if(locale_path && !g_file_test(locale_path, G_FILE_TEST_EXISTS))
+	{
+		g_free(locale_path);
+		locale_path = NULL;
+	}
+
+	/* try to get path from project */
+	if (!locale_path && geany_data->app->project)
+	{
+		gchar *utf8_path = NULL;
+		utf8_path = get_project_base_path();
+		locale_path = utils_get_locale_from_utf8(utf8_path);
+		g_free(utf8_path);
+	}
+
+	/* Make sure path exists */
+	if(locale_path && !g_file_test(locale_path, G_FILE_TEST_EXISTS))
+	{
+		g_free(locale_path);
+		locale_path = NULL;
+	}
+
+	return locale_path;
+}
+
+
+/* returns parent folder of path from get_full_path_for_selection()
+ *   if unable to find parent folder, returns home folder
+ *   otherwise, returns NULL
+ * return path is in locale encoding */
+static gchar *get_folder_for_selection(void)
+{
+	gchar *locale_path;
+	locale_path = get_fullpath_for_selection();
+
+	/* Get parent folder if regular file */
+	if (locale_path && g_file_test(locale_path, G_FILE_TEST_IS_REGULAR))
+	{
+		gchar *folder;
+		folder = g_path_get_dirname(locale_path);
+		if (folder && folder == '.')
+		{
+			g_free(folder);
+			folder = NULL;
+		}
+		if (folder && g_file_test(folder, G_FILE_TEST_IS_DIR))
+		{
+			g_free(locale_path);
+			locale_path = folder;
+		}
+		else
+		{
+			g_free(folder);
+			g_free(locale_path);
+			locale_path = NULL;
+		}
+	}
+
+	/* Try home folder */
+	if (!locale_path)
+		locale_path = g_strdup(g_get_home_dir());
+
+	/* Make sure path exists */
+	if(locale_path && !g_file_test(locale_path, G_FILE_TEST_IS_DIR))
+	{
+		g_free(locale_path);
+		locale_path = NULL;
+	}
+
+	return locale_path;
+}
```

Comment:
I think both get_fullpath_for_selection() and get_folder_for_selection() are an overkill. I think you could just simply get the path like in on_delete() and then test whether the resulting path is a file or directory, remove the file part if it's a file and use the resulting path to open file manager or terminal. I don't think any super-strict checks are necessary here - in the worst case the terminal or the file manager get opened at a wrong path.